diff -ur onig-5.7.0.orig/regint.h onig-5.7.0/regint.h
--- onig-5.7.0.orig/regint.h	2007-06-12 13:37:26.000000000 +0900
+++ onig-5.7.0/regint.h	2007-06-12 13:38:07.000000000 +0900
@@ -80,7 +80,6 @@
 
 /* */
 /* escape other system UChar definition */
-#include "config.h"
 #ifdef ONIG_ESCAPE_UCHAR_COLLISION
 #undef ONIG_ESCAPE_UCHAR_COLLISION
 #endif
@@ -140,7 +139,7 @@
 #else
 #define xalloca     alloca
 #endif
-
+#include "config.h"
 #if defined(USE_RECOMPILE_API) && defined(USE_MULTI_THREAD_SYSTEM)
 #define ONIG_STATE_INC(reg) (reg)->state++
 #define ONIG_STATE_DEC(reg) (reg)->state--
