Index: stlport/stl/_tree.h
===================================================================
--- stlport/stl/_tree.h	(revision 3964)
+++ stlport/stl/_tree.h	(working copy)
@@ -262,18 +262,18 @@
   _Compare _M_key_compare;
 
   _Link_type& _M_root() const 
-    { return (_Link_type&) this->_M_header._M_data->_M_parent; }
+    { return (_Link_type&) *this->_M_header._M_data->_M_parent; }
   _Link_type& _M_leftmost() const 
-    { return (_Link_type&) this->_M_header._M_data->_M_left; }
+    { return (_Link_type&) *this->_M_header._M_data->_M_left; }
   _Link_type& _M_rightmost() const 
-    { return (_Link_type&) this->_M_header._M_data->_M_right; }
+    { return (_Link_type&) *this->_M_header._M_data->_M_right; }
 
   static _Link_type& _STLP_CALL _S_left(_Link_type __x)
-    { return (_Link_type&)(__x->_M_left); }
+    { return (_Link_type&)(*__x->_M_left); }
   static _Link_type& _STLP_CALL _S_right(_Link_type __x)
-    { return (_Link_type&)(__x->_M_right); }
+    { return (_Link_type&)(*__x->_M_right); }
   static _Link_type& _STLP_CALL _S_parent(_Link_type __x)
-    { return (_Link_type&)(__x->_M_parent); }
+    { return (_Link_type&)(*__x->_M_parent); }
   static reference  _STLP_CALL _S_value(_Link_type __x)
     { return __x->_M_value_field; }
   static const _Key& _STLP_CALL _S_key(_Link_type __x)
